[quant][pt2e] Fix handling for SharedQuantizationSpec

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106922

Summary:
Previously if we have:
```
conv1 -> cat
conv2  /
```
and configure output of conv1/conv2 to be int8 quantized, and cat also int8 quantized and with shared inputs,
it will not produce expected results (input of cat will not be shared)

The problem is that there is some missing checks when inserting observers for input for cat

This PR fixes the problem.

Fixes: https://github.com/pytorch/pytorch/issues/106760
Test Plan:
python tes/test_quantization.py TestQuantzePT2E.test_shared_qspec

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e] Fix handling for SharedQuantizationSpec

Summary:
Previously if we have:
```
conv1 -> cat
conv2  /
```
and configure output of conv1/conv2 to be int8 quantized, and cat also int8 quantized and with shared inputs,
it will not produce expected results (input of cat will not be shared)

The problem is that there is some missing checks when inserting observers for input for cat

This PR fixes the problem.

Test Plan:
python tes/test_quantization.py TestQuantzePT2E.test_shared_qspec

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Fix handling for SharedQuantizationSpec"


Summary:
Previously if we have:
```
conv1 -> cat
conv2  /
```
and configure output of conv1/conv2 to be int8 quantized, and cat also int8 quantized and with shared inputs,
it will not produce expected results (input of cat will not be shared)

The problem is that there is some missing checks when inserting observers for input for cat

This PR fixes the problem.

Test Plan:
python tes/test_quantization.py TestQuantzePT2E.test_shared_qspec

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/pt2e/prepare.py:
```diff
@@ -62,6 +64,9 @@ def _maybe_insert_input_observer_for_arg_or_kwarg(
     arg_as_output_target_dtype, arg_as_output_target_is_dynamic = _get_dtype_and_is_dynamic(arg_as_output_act_obs_or_fq)
 
+    input_qspec_map = quantization_annotation.input_qspec_map
+    input_arg_qspec = _get_qspec_for_arg(arg, input_qspec_map, named_modules)
```
This is used only when you replacing already existing observer code. Maybe move it closer to the use?

jerryzh168 committed:
Update on "[quant][pt2e] Fix handling for SharedQuantizationSpec"


Summary:
Previously if we have:
```
conv1 -> cat
conv2  /
```
and configure output of conv1/conv2 to be int8 quantized, and cat also int8 quantized and with shared inputs,
it will not produce expected results (input of cat will not be shared)

The problem is that there is some missing checks when inserting observers for input for cat

This PR fixes the problem.

Fixes: https://github.com/pytorch/pytorch/issues/106760
Test Plan:
python tes/test_quantization.py TestQuantzePT2E.test_shared_qspec

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/pt2e/prepare.py:
```diff
@@ -71,7 +76,16 @@ def _maybe_insert_input_observer_for_arg_or_kwarg(
             assert observed_arg in obs_or_fq_map, \
                 f"can't refer to a node that does not have observer/fake_quant inserted yet: {observed_arg}"
-            arg_as_input_act_obs_or_fq = obs_or_fq_map[observed_arg]
+            if isinstance(input_arg_qspec, SharedQuantizationSpec):
```
So if this is shared qspec, then wont you already be sharing observer? What is the observer replcaement being done here? Is it replacing the output observer of previous node? Not clear.

jerryzh168 replied:
yeah this is replacing the output observer from previous op

what do you mean by it's already sharing observer? user configures the input of the op to share with some other things, this is in the process of implementing it

kimishpatel replied:
Basically if you have
conv1 (output=qint8).    conv2 (qint8)
  |                                        |
obs1                             obs2
 |                                     |
cat (wants shared obs)

then each conv1 and conv2 will already have their observers inserted when you come to cat. So I am asking will cat replcace obs1 with obs2? Or will you have some like

conv1 (output=qint8).    conv2 (qint8)
  |                                        |
obs1                             obs2
 |                                     |
 |                                     |
shared_obs3       shared_obs3
  |                             |
cat (wants shared obs)

It wasnt clear to me.

kimishpatel replied:
my formatting is all messed up

kimishpatel replied:
But basically if conv1 and conv2 have their observers inserted already and the output didnt have "shared quant spec", then here you are forcing it to share? That part wasnt clear to me.

kimishpatel replied:
and then my question would be: is it better to do this as preprocess step on annotations, that figures out that conv1 and conv2's output is going into cat which wants shared qspec, then let me update conv1 and conv2's output_qspec to be shared?

jerryzh168 replied:
> But basically if conv1 and conv2 have their observers inserted already and the output didnt have "shared quant spec", then here you are forcing it to share? That part wasnt clear to me.

Yes

jerryzh168 replied:
> and then my question would be: is it better to do this as preprocess step on annotations, that figures out that conv1 and conv2's output is going into cat which wants shared qspec, then let me update conv1 and conv2's output_qspec to be shared?

hmmm, I think that would require additional analysis beforehand that would require more code, what is the benefit?

kimishpatel replied:
I think both approaches are fine. was just pointing out alternatives.

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1003,6 +1017,133 @@ def get_supported_operators(cls) -> List[OperatorConfig]:
+                input1 = inputs[1]
+                assert input0.op == "call_module"
+                assert input1.op == "call_module"
+                obs_ins0 = getattr(m, input0.target)
+                obs_ins1 = getattr(m, input1.target)
```
Shouldnt you also ensure that output observers of conv1 and conv2 are also shared/same?

jerryzh168 replied:
oh sure

kimishpatel reviewed (changes requested):
left some comments

jerryzh168 committed:
Update on "[quant][pt2e] Fix handling for SharedQuantizationSpec"


Summary:
Previously if we have:
```
conv1 -> cat
conv2  /
```
and configure output of conv1/conv2 to be int8 quantized, and cat also int8 quantized and with shared inputs,
it will not produce expected results (input of cat will not be shared)

The problem is that there is some missing checks when inserting observers for input for cat

This PR fixes the problem.

Fixes: https://github.com/pytorch/pytorch/issues/106760
Test Plan:
python tes/test_quantization.py TestQuantzePT2E.test_shared_qspec

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
